### PR TITLE
feat: Allow binary artifact

### PR DESCRIPTION
**Pull request by atsteffen**

### CODE REVIEW

**Comment by amunra** [comment]
This is fantastic!
I'll take a proper look on Monday.

I do wonder if it makes sense to have separate:
```
<copyLibTo>...path1...</copyLibTo>
<copyBinsTo>...path2...</copyBinsTo>
```

Just in case a Rust crate has both libs and bins.

**Comment by atsteffen** [comment]
Good point! For the general case, this PR is definitely incomplete. There can be 1 library output matching the project name, and multiple named binaries, only one of which might match the project name.  I'll put together a test case that has a library and multiple binaries, and approach this from a more general perspective.

**Comment by amunra** [comment]
The PR is great, if not for this detail. 

I think the code needs to parse the toml to find out what needs to be extracted, be it a lib or other and then copy to type-specific paths.

If this is more work than you signed up for I'm happy to get started and ping you once the code  has the toml parsing implemented. It will take a few days though as I'm working on something else today and tomorrow. Up to you.

**Comment by atsteffen** [comment]
I'm not in a hurry on this, but I can continue the implementation process on this thread.

**Comment by atsteffen** [comment]
Initial Cargo.toml parsing added. More work is needed on acceptance tests and adding crate_type parsing.

Specific feedback would be helpful before proceeding:
1) Is it ok for the rust-maven-example project to be the home of acceptance tests (where cargo is run using the plugin and asserts are made on those outputs)?  This places the tests slightly more distant to the plugin code, and dirties the example project in case it was intended as a simple template project. On the other hand, acceptance tests make great examples, so it's perhaps not too out of place.
2) Should the <copyTo> try to eventually use --out-dir?  In that case, I would stick to a single <copyTo> to stay in inline with the cargo arg, and to keep the implementation simpler. maven-resources-plugin or build scripts can be used by clients to customize file destinations further.

**Comment by amunra** [comment]
> Initial Cargo.toml parsing added. More work is needed on acceptance tests and adding crate_type parsing.
> 
> Specific feedback would be helpful before proceeding:
> 1) Is it ok for the rust-maven-example project to be the home of acceptance tests (where cargo is run using the plugin and asserts are made on those outputs)?  This places the tests slightly more distant to the plugin code, and dirties the example project in case it was intended as a simple template project. On the other hand, acceptance tests make great examples, so it's perhaps not too out of place.
> 2) Should the "copyTo" arg try to eventually use --out-dir?  In that case, I would stick to a single "copyTo" to stay in inline with the cargo arg, and to keep the implementation simpler. maven-resources-plugin or build scripts can be used by clients to customize file destinations further.

Looking really good at first glance: I'll do a proper review of this tomorrow.

1) Yes.
2) Probably: You make a good point, but I need to think about it further in case I'm missing something: It's certainly easier to just have one single `<copyTo>` path.

Reading `--out-dir`.. who knows. Until its in stable it's difficult to know what the final behavior will be.

File: rust-maven-example/src/main/java/io/questdb/example/rust/Main.java [hunk 1]
```diff
@@ -29,7 +29,7 @@
 public class Main {
     public static native String reversedString(String str);
 
-    static {
+    public static void main(String[] args) {
```
Comment (amunra) [inline rust-maven-example/src/main/java/io/questdb/example/rust/Main.java:32]: I think this should stay in a `static` block as it's generally how JNI libs are loaded.
No need to change this code.

File: rust-maven-example/src/main/rust/str-reverse-binary/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+[package]
+name = "str-reverse-binary"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+exitcode = "1.1.2"
```
Comment (amunra) [inline rust-maven-example/src/main/rust/str-reverse-binary/Cargo.toml:7]: This dependency can be removed. Just slows the build down.

File: rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+use std::env;
+use std::process::exit;
```
Comment (amunra) [inline rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs:26]: `use std::process::ExitCode;`

File: rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+fn main() {
```
Comment (amunra) [inline rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs:28]: `fn main() -> ExitCode {`

File: rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs (hunk 3 above)
Comment (amunra) [reply rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs:26]: Return the exit code instead of calling `exit()`.

File: rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    let args: Vec<String> = env::args().collect();
+
+    match args.len() {
+        2 => {
+            let reversed: String = args[1].chars().rev().collect();
+            println!("{}", reversed);
+            exit(exitcode::OK);
```
Comment (amunra) [inline rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs:35]: `ExitCode::SUCCESS`

File: rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        },
+        _ => {
+            eprintln!("Error: Must provide a single string argument.");
+            exit(exitcode::CONFIG);
```
Comment (amunra) [inline rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs:39]: `ExitCode::FAILURE`

File: rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs (hunk 3 above)
Comment (amunra) [reply rust-maven-example/src/main/rust/str-reverse-binary/src/main.rs:26]: https://stackoverflow.com/a/54174631/4319200

File: rust-maven-example/src/test/java/io/questdb/example/rust/LibTest.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+import io.questdb.jar.jni.LibInfo;
+
+public class LibTest {
+
+    @Test
+    public void testLibrary() {
+        File libraryFile = new File("target/rust-maven-plugin/str-reverse/release/", new LibInfo("str_reverse").getFullName());
+
+        System.load(libraryFile.getAbsolutePath());
```
Comment (amunra) [inline rust-maven-example/src/test/java/io/questdb/example/rust/LibTest.java:41]: This shouldn't be necessary once the loading code is moved to a `static` block.

Did you encounter issues?

File: rust-maven-plugin/src/main/java/io/questdb/maven/rust/CargoBuildMojo.java [hunk 8]
```diff
@@ -161,8 +159,8 @@ private String getName() {
         return components[components.length - 1];
     }
 
-    private File getTargetDir() {
-        return new File(new File(new File(project.getBuild().getDirectory()), "rust-maven-plugin"), getName());
+    private File getTargetDir(String name) {
+        return new File(new File(new File(project.getBuild().getDirectory()), "rust-maven-plugin"), name);
```
Comment (amunra) [inline rust-maven-plugin/src/main/java/io/questdb/maven/rust/CargoBuildMojo.java:163]: Revert this change: There's no need to parameterize this as the only thing that's ever  passed in is `getName()`. Simpler code is better.

File: rust-maven-plugin/src/main/java/io/questdb/maven/rust/CargoBuildMojo.java [hunk 9]
```diff
@@ -230,50 +228,15 @@ private File getCopyToDir() throws MojoExecutionException {
...
-        getLog().info("Copied " + Shlex.quote(getArtifactPath.getName()));
-    }
-
     @Override
     public void execute() throws MojoExecutionException {
         List<String> args = new ArrayList<>();
         args.add("build");
 
+        final String name = getName();
+        final File targetDir = getTargetDir(name);
```
Comment (amunra) [inline rust-maven-plugin/src/main/java/io/questdb/maven/rust/CargoBuildMojo.java:237]: Revert change.

File: rust-maven-plugin/src/main/java/io/questdb/maven/rust/CargoBuildMojo.java [hunk 10]
```diff
@@ -230,50 +228,15 @@ private File getCopyToDir() throws MojoExecutionException {
...
         args.add("--target-dir");
-        args.add(getTargetDir().getAbsolutePath());
+        args.add(targetDir.getAbsolutePath());
```
Comment (amunra) [inline rust-maven-plugin/src/main/java/io/questdb/maven/rust/CargoBuildMojo.java:239]: Revert change.

**Review by amunra — CHANGES_REQUESTED** [review]
Some very minor tweaks requested.
I very much agree with your general approach.
Thanks!

PS. You'll need to also sign the contributor agreement.

File: rust-maven-example/src/test/java/io/questdb/example/rust/LibTest.java (hunk 7 above)
Comment (atsteffen) [reply rust-maven-example/src/test/java/io/questdb/example/rust/LibTest.java:41]: It might have been a fluke of my IDE. I thought Class.getResourceAsStream wasn't working in the test, but checking it again, it seems to work fine.

File: rust-maven-plugin/src/main/java/io/questdb/maven/rust/CargoBuildMojo.java (hunk 8 above)
Comment (atsteffen) [reply rust-maven-plugin/src/main/java/io/questdb/maven/rust/CargoBuildMojo.java:163]: It was just a micro optimization as getName() does a string split and gets called twice now that ManualOutputRedirector uses it.  But I agree simplicity > micro optimizations in this case.

**Comment by atsteffen** [comment]
Thanks for the feedback and good tweaks!  We are considering a partial transition from c++ to rust, and I don't have a lot of rust/cargo experience (yet), so this is a good learning exercise for me.  Contributing high coverage test cases to this plugin project has some significant learning value for me.

**Comment by atsteffen** [comment]
> PS. You'll need to also sign the contributor agreement.

It says it is signed but pending.

**Review by amunra — APPROVED** [review]
